### PR TITLE
Notify users in the CLI when a new version of Rill is available

### PR DESCRIPTION
**Pull request by rakeshsharma14317**
Currently we lookup for the new release once in 24h (and cache it locally in .rill/state.yaml)  and show the message every time user runs the cli command with update notification.

### CODE REVIEW

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Unfortunately, this adds a new network request to check the version number before every CLI command – which makes each CLI call one network roundtrip slower. 

Instead, this should be implemented by:
1. The admin server setting a custom response header containing the latest version number when the calling user agent is `rill-cli` 
2. The CLI parsing it for every RPC (using a client-side interceptor) and checking it against its own version number, then printing a nudge to upgrade if not

**Comment by hamilton** [comment]
@rakeshsharma14317 @begelundmuller is there a way to expose this information in the frontend through an API as well?

**Comment by begelundmuller** [comment]
> @rakeshsharma14317 @begelundmuller is there a way to expose this information in the frontend through an API as well?

@hamilton, yeah, we could run the check and return the result (current and highest known version number) on `/local/version` (similar to `/local/config`). It should be a separate endpoint since it resolves slowly and `/local/config` holds up the first render.

@rakeshsharma14317, can you add that endpoint to this PR?

File: cli/pkg/local/app.go [hunk 1]
```diff
@@ -308,6 +309,17 @@ func (a *App) Serve(httpPort, grpcPort int, enableUI, openBrowser, readonly bool
...
+	// Get the latest version available
+	latestVersion, err := update.LatestVersion(gctx)
+	if err != nil {
+		a.Logger.Warnf("error finding latest version: %v", err)
+	}
+
+	versionInfo := &versionInfo{
+		CurrentVersion: a.Version.Number,
+		LatestVersion:  latestVersion.Version,
+	}
```
Comment (begelundmuller) [inline cli/pkg/local/app.go:321]: Run this inside the handler func in `versionHandler` so it doesn't add latency to the startup

File: cli/pkg/local/app.go [hunk 2]
```diff
@@ -381,6 +394,11 @@ func (a *App) pollServer(ctx context.Context, httpPort int, openOnHealthy bool)
 	}
 }
 
+type versionInfo struct {
+	CurrentVersion string `json:"current_version"`
+	LatestVersion  string `json:"latest_version"`
+}
```
Comment (begelundmuller) [inline cli/pkg/local/app.go:400]: nit: Move to right after the `versionHandler` func

File: cli/pkg/update/update.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+	"github.com/fatih/color"
+	"github.com/hashicorp/go-version"
+	"github.com/rilldata/rill/cli/pkg/dotrill"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	StateFilename       = "state.yaml"
+	VersionInfoStateKey = "version_info"
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:20]: These should be defined in `dotrill`

File: cli/pkg/update/update.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		return err
+	}
+
+	err = dotrill.Set(StateFilename, VersionInfoStateKey, string(content))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:209]: It shouldn't serialize yaml here, it should just set two keys: `latest_version` and `latest_version_checked_at`

File: cli/pkg/update/update.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	defer resp.Body.Close()
+
+	out, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	success := resp.StatusCode >= 200 && resp.StatusCode < 300
+	if !success {
+		return nil, fmt.Errorf("error fetching latest release: %v", string(out))
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:160]: Use `%s`

File: cli/pkg/update/update.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	return &UpdateInfo{
+		Update: false,
+		Message: fmt.Sprintf("Latest version (%s) is less than or equal to current build version (%s)",
+			info.Version, currentVersion),
+		ReleaseInfo: info,
+	}, nil
+}
+
+// This will fetch the latest version available for rill on github releases
+func LatestVersion(ctx context.Context) (*ReleaseInfo, error) {
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:126]: Should be internal: `fetchLatestVersion`

File: cli/pkg/update/update.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+	Message     string
+	ReleaseInfo *ReleaseInfo
+}
+
+// ReleaseInfo stores information about a release
+type ReleaseInfo struct {
+	Version     string    `json:"tag_name"`
+	URL         string    `json:"html_url"`
+	PublishedAt time.Time `json:"published_at"`
+}
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:35]: Make it `githubReleaseInfo` and define it right above the function that returns it

File: cli/pkg/update/update.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+		fmt.Printf("\n%s %s → %s\n\n",
+			color.YellowString("A new version of rill is available:"),
+			color.CyanString(currentVersion),
+			color.CyanString(updateInfo.ReleaseInfo.Version))
+	}
+
+	return nil
+}
+
+func checkVersion(ctx context.Context, currentVersion string) (*UpdateInfo, error) {
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:58]: Replace the version getting and checking logic here with a `LatestVersion` function that returns the cached version number or calls `fetchLatestVersion` if cache is outdated.

Comment (begelundmuller) [reply cli/pkg/update/update.go:58]: It should just return the newest version number, nothing else

Comment (begelundmuller) [reply cli/pkg/update/update.go:58]: Then move the version checking logic into `CheckVersion`

**Review by begelundmuller — CHANGES_REQUESTED** [review]
The code is a little hard to follow here – in general, only use `Uppercase` structs, consts, and functions names if they are accessed outside the package.

File: cli/pkg/dotrill/dotrill.go [hunk 9]
```diff
@@ -122,6 +124,22 @@ func SetAccessToken(token string) error {
...
+	return Get(StateFilename, VersionUpdatedAtKey)
+}
+
+func SetVersionUpdatedAt(updatedAt string) error {
+	return Set(StateFilename, VersionUpdatedAtKey, updatedAt)
+}
+
+func SetVersion(version string) error {
+	return Set(StateFilename, VersionKey, version)
+}
```
Comment (begelundmuller) [inline cli/pkg/dotrill/dotrill.go:141]: Order related functions together, with setter first, like the other ones above

File: cli/pkg/local/app.go [hunk 10]
```diff
@@ -434,6 +436,39 @@ func (a *App) infoHandler(info *localInfo) http.Handler {
...
+		a.Logger.Warnf("error finding latest version: %v", err)
+	}
+
+	inf := &versionInfo{
+		CurrentVersion: a.Version.Number,
+		LatestVersion:  latestVersion,
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		data, err := json.Marshal(inf)
```
Comment (begelundmuller) [inline cli/pkg/local/app.go:453]: 1. Move this into the handler, so it doesn't block the initial server startup (it will be cached inside `update.LatestVersion` anyway)
2. Use `r.Context()` and don't accept `ctx` as parameter

File: cli/pkg/update/update.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+		return "", err
+	}
+
+	err = dotrill.SetVersion(latestVersion)
+	if err != nil {
+		return "", err
+	}
+
+	return latestVersion, nil
+}
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:68]: Move the calls to `SetXXX` into `LatestVersion`, and remove this function (just call `LatestVersion` directly from `CheckVersion`

File: cli/pkg/update/update.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+			info, err := fetchLatestVersion(ctx)
+			if err != nil {
+				return "", err
+			}
+
+			return info.Version, nil
+		}
+
+		return cachedVersion, nil
+	}
```
Comment (begelundmuller) [inline cli/pkg/update/update.go:99]: 1. Change this condition. If `time.Since` is less than the TTL, return the cached version. Else, don't return anything, so the fetch on line 102 is triggered.
2. Also, make the TTL a const at the top of the file – like `const versionCheckTTL = 24*time.Hour`

**Review by begelundmuller — CHANGES_REQUESTED** [review]

**Review by begelundmuller — APPROVED** [review]
